RoR - Private events associations project

guirecordon opened the pull request:
### Collaborative project - Request for review
**We did the following for this project and now feel ready to request for a review. We: **

- created a User, an Event, and an EventManager model
- wrote the associations for them with their foreign keys and many-to-many relationships
- wrote validations for the users and event creators to be able to log in and only have access to their respective actions
- created a way for the users to be able to accept their invitations once they were made by another user. We did that by adding a 'confirmation' column to our EventManager table and only displaying the events once they were marked as true. They were all set to false in the beginning.
- we wrote integration tests and unit tests for authentication and models

thank you.

Jmagero committed:
Add user model

Jmagero committed:
Add column username & email

Jmagero committed:
Add users controller

Jmagero committed:
Add simple signin function

guirecordon committed:
Create event model and migrations

Jmagero committed:
Add events controller

guirecordon committed:
Build associations for event creator

Jmagero committed:
Add description to event model

guirecordon committed:
Change events create method to use build

guirecordon committed:
Attempt to display username on event show

Jmagero committed:
Add  events show page

Jmagero committed:
Add event index page

Jmagero committed:
Add association between event attendee & even

Jmagero committed:
Add association between attendee & event

Jmagero committed:
Add associations between attendee & event

guirecordon committed:
Finish event_managers associations

Jmagero committed:
Add attendees to events show page

Jmagero committed:
Add list of events to user's show page

Jmagero committed:
Add column reset to Fixcolumn migration

Jmagero committed:
Separate events on user show

guirecordon committed:
Add log-in out to navbar

guirecordon committed:
Add navigation links

Jmagero committed:
Add event_manager controller

Jmagero committed:
Fix merge conflict

Jmagero committed:
Fix merge conflict

guirecordon committed:
Attempt to activate invites

Jmagero committed:
Add create method to eventmanager

guirecordon committed:
New attempt at invitations

Jmagero committed:
Add invitation form

guirecordon committed:
Add confirmation column to event_manager

guirecordon committed:
Add invite confirmation view and flow

Jmagero committed:
Add invited method

guirecordon committed:
Fix create method in events_controller

Jmagero committed:
Add rspec

guirecordon committed:
Install coverage gem for test

guirecordon committed:
Create tests for event associations

Jmagero committed:
Add user validation test

Jmagero committed:
Add user validation test

guirecordon committed:
attempt to write rspec tests

guirecordon committed:
Add test sign up user

Jmagero committed:
Tried to add tests for event_manager

guirecordon committed:
Change event manager spec

guirecordon committed:
Attempt to select dropdown menu items

Jmagero committed:
Attempt to test event_manager

Jmagero committed:
Add Create event link

guirecordon committed:
Change select in event_manager_spec

guirecordon committed:
Pass tests

guirecordon committed:
Update README file

Jmagero committed:
set up rubocop

Jmagero committed:
Set up linter file

Jmagero committed:
Fix rubocop issues

Jmagero committed:
Add stylelint setup

Jmagero committed:
Attempt to install npm

guirecordon committed:
Run stylint and fix errors

mosesogwo reviewed (changes requested):
**Hello!:wave: Good start on this project:+1: Here are your required improvements:**

:point_right: Event's index page should display past and upcoming events lists. This isn't working right now. Events are not being displayed.

![](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png)

:point_right: The event's show page displays the list of attendants. Attendants are not  being displayed currently.

![](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/events_attendees_requirement.png)

:point_right:  The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming. The events are not visible on the page.

:point_right: Event model has an event date attribute. So you have to add the date attribute to your event creation form.

![](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png)

:point_right: Events controller index action should list past and upcoming events. This grouping should be made in the controller and not views.

;point_right: Users controller show action should have the following:

```
lists events created by user, i.e. @created_events =  current_user.events (screenshot from Odin)

 lists events previoulsy attendented by user using event model scope, i.e. @past_events = current_user.attended_events.previous (screenshot from Odin)

 lists upcoming events previously attended by user using event model scope, i.e. @upcoming_events = current_user.attended_events.upcoming (screenshot from Odin)

```
:point_right: Event management feature spec should cover the entire event management feature. You should send an invite in your spec and verify that it's working.

:point_right: Your tests are failing. Please fix them.

_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested:exclamation:

Jmagero committed:
Add drop down menu test to event-manager

Jmagero committed:
Merge branch 'feature' of https://github.com/Jmagero/private-events into feature

Jmagero committed:
Add attendees to event's show page

guirecordon committed:
Remove T/False conditional user controller

guirecordon committed:
Change destination after signup to login

Jmagero committed:
Add events created by user on the user's show page

Jmagero committed:
Create methods for upcoming and previous events

guirecordon committed:
Attempt to fix EventManager test

Jmagero committed:
Add event_manager tests

guirecordon committed:
Run rubocop

guirecordon commented:
Thank you, Moses.
We worked on the changes you requested. 

- The upcoming and past events should be working now
- Attendants of the events should also be displayed (we had to change the conditional for the invite accept in the users controller -- confirmed.zero?)
- The date attribute is working and we also added a couple of links, etc for better user experience.
- Finally, we seem to have been able to get the EventManager RSpec test to pass, but that was a tricky one. So we'd love to get some feedback on it.

We're requesting a new review.
Thank you

maelfosso commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,49 @@
+
+<ul>
+<strong>Pending Invitations:</strong>
+    <% if @pending_invitations.size == 0 %>
+      <p><em> You got no pending invitations </em> </p>
+    <% else %>
+      <% @pending_invitations.each do |e| %>
+        <li>
+          <p>Invitation ID: <%= link_to e.attended_event_id, event_path(e.attended_event_id) %></p>
+          <p><%#= e.name %></p>
+          <p><%= link_to "Accept Invitation", event_manager_path(e) %></p>
+        </li>
+      <% end %> 
+    <% end %>
```
Kindly remove login from views, move it in view helpers

maelfosso commented on app/views/layouts/application.html.erb:
```diff
@@ -0,0 +1,29 @@
+
+  <body>
+
+    <% if current_user %>
+      <div class="nav-bar d-flex">
+        <p>Logged in as <%= current_user.username %></p>
+        <%= link_to "Log Out", session_path(current_user), method: :delete %>
+      </div>  
+    <% else %>
+      <div class="nav-bar d-flex">
+        <%= link_to "Sign up", new_user_path %> or
+        <%= link_to "Log In", new_session_path %>
+      </div>
+    <% end %>
```
Move logic into view helpers

maelfosso commented on app/views/events/index.html.erb:
```diff
@@ -0,0 +1,24 @@
+<h1> Events</h1>
+
+<% if current_user %>
+  <%= link_to "Create New Event", new_event_path %>
+<% end %> 
```
Move logic into view helpers

maelfosso commented on app/views/events/show.html.erb:
```diff
@@ -0,0 +1,18 @@
+<h1><%= @event.name %></h1>
+<p>Event created by <%= @event.creator.username %></p>
+<h2>Invited Guests</h2>
+    <% if @attendants.size == 0 %>
+        <p><em>No guests have been invited to this event so far.</em></p>
+    <% else %>
+        <% @attendants.each do |e| %>
+        <%= e.username %><br>
+    <% end %>
```
Move logic into view helpers

maelfosso commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  def show
+    @user = User.find(params[:id])
+    @pending_invitations = []
+    @upcoming_events = current_user.attended_events.where('date >= ?', DateTime.now)
+    @previous_events = current_user.attended_events.where('date <= ?', DateTime.now)
```
No longer need to add request. You already have the scope `Event`.
Directly used
```rb
@upcoming_events = current_user.attended_events.upcoming
@previous_events = current_user.attended_events.past
```

maelfosso commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,29 @@
+    @upcoming_events = current_user.attended_events.where('date >= ?', DateTime.now)
+    @previous_events = current_user.attended_events.where('date <= ?', DateTime.now)
+
+    @user.event_managers.each do |invite|
+      confirmed = invite.confirmation
+      @pending_invitations << invite if confirmed.zero?
+    end
```
No need to confirm the invitation!

maelfosso reviewed (changes requested):
# Changes required :red_circle: 

Hello,

Thank you for applying the suggested changes
However, there are still some issues 

1. There are no information concerning the attended users, as required [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/events_attendees_requirement.png).
![Screenshot at 2020-09-14 19:09:50](https://user-images.githubusercontent.com/8848104/93122142-e5a0e480-f6bd-11ea-8490-374abb8b0f4f.png)
But if you take a look at the database, you will notice that there are invited users
```rb
2.7.1 :004 > EventManager.all
  EventManager Load (0.3ms)  SELECT  "event_managers".* FROM "event_managers" LIMIT ?  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<EventManager id: 1, attendee_id: 1, attended_event_id: 1, created_at: "2020-09-14 18:02:40", updated_at: "2020-09-14 18:02:40", confirmation: 0>, #<EventManager id: 2, attendee_id: 1, attended_event_id: 4, created_at: "2020-09-14 18:04:00", updated_at: "2020-09-14 18:04:00", confirmation: 0>]> 
2.7.1 :005 > Event.all
  Event Load (0.3ms)  SELECT  "events".* FROM "events" LIMIT ?  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Event id: 1, name: "Music go", date: "2020-09-17 00:00:00", created_at: "2020-09-14 18:02:31", updated_at: "2020-09-14 18:02:31", creator_id: 1, description: "Yeah you are OK. https://github.com/emasdev/privat...">, #<Event id: 2, name: "Music go", date: "2020-09-01 00:00:00", created_at: "2020-09-14 18:02:50", updated_at: "2020-09-14 18:02:50", creator_id: 1, description: "Jour de ca ah bon!!!">, #<Event id: 3, name: "Pourrational high", date: "2020-09-30 00:00:00", created_at: "2020-09-14 18:03:22", updated_at: "2020-09-14 18:03:22", creator_id: 1, description: "Yeah you are OK. https://github.com/emasdev/privat...">, #<Event id: 4, name: "Reviewing", date: "2020-10-03 00:00:00", created_at: "2020-09-14 18:03:51", updated_at: "2020-09-14 18:03:51", creator_id: 2, description: "Read it correctly youre chaongre ;w;oijwe">]> 
```

2- There is no link to access to the user show page which is requested. [Check it here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png)

3- The model `AddCreatorToEvent`  is never used. Can you remove it ?

4- Read the comments below

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

guirecordon committed:
Add partials to views and application

Jmagero committed:
Add partials to events & users pages

guirecordon committed:
Remove loggin from views to helpers

Jmagero committed:
Add concat to user_log_in helper method

guirecordon committed:
Finish removal of logic from views

guirecordon committed:
Run rubocop

guirecordon commented:
Thank you, Moses.

We removed all the logic from views and moved them to helpers; added the links you requested and made the other changes.

Thank you.

Dandush03 reviewed (approved):
# Status: :rocket: Approved :rocket: :smile: 
# Keep it up
# :computer: Happy Coding :computer:
